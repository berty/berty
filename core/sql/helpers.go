package sql

import (
	"errors"

	"github.com/jinzhu/gorm"

	"github.com/berty/berty/core/entity"
)

func ContactByID(db *gorm.DB, id string) (*entity.Contact, error) {
	var contact entity.Contact
	return &contact, db.First(&contact, "ID = ?", id).Error
}

// FindContact tries different approaches to get a full contact based on a partial one
func FindContact(db *gorm.DB, input *entity.Contact) (*entity.Contact, error) {
	if input.ID != "" {
		return ContactByID(db, input.ID)
	}
	// FIXME: support looking-up by sigchain
	return nil, errors.New("not enough information to search contact")
}
