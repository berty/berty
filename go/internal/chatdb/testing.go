package chatdb

import (
	"testing"

	"github.com/jinzhu/gorm"
	"go.uber.org/zap"
)

// TestingSqliteDB returns a configured chat gorm database.
func TestingSqliteDB(t *testing.T, logger *zap.Logger) *gorm.DB {
	t.Helper()

	db, err := gorm.Open("sqlite3", ":memory:")
	if err != nil {
		t.Fatalf("failed to initialize in-memory sqlite srever: %v", err)
	}

	db, err = Init(db, logger)
	if err != nil {
		t.Fatalf("failed to configure database: %v", err)
	}

	err = Migrate(db, false, logger)
	if err != nil {
		t.Fatalf("failed to run migrations: %v", err)
	}

	return db
}
