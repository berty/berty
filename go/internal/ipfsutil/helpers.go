package ipfsutil

import (
	"context"
	"errors"
	"fmt"
	"io"
	"sync"
	"time"

	"github.com/libp2p/go-libp2p-core/network"
	"github.com/libp2p/go-libp2p-core/peer"
	ma "github.com/multiformats/go-multiaddr"
	madns "github.com/multiformats/go-multiaddr-dns"
	"go.uber.org/multierr"
	"go.uber.org/zap"
	"go.uber.org/zap/zapcore"
)

func ParseAndResolveIpfsAddr(ctx context.Context, addr string) (*peer.AddrInfo, error) {
	maddr, err := ma.NewMultiaddr(addr)
	if err != nil {
		return nil, err
	}

	if !madns.Matches(maddr) {
		return peer.AddrInfoFromP2pAddr(maddr)
	}

	addrs, err := madns.Resolve(ctx, maddr)
	if err != nil {
		return nil, err
	}
	if len(addrs) == 0 {
		return nil, errors.New("fail to resolve the multiaddr:" + maddr.String())
	}

	var info peer.AddrInfo
	for _, addr := range addrs {
		taddr, id := peer.SplitAddr(addr)
		if id == "" {
			// not an ipfs addr, skipping.
			continue
		}
		switch info.ID {
		case "":
			info.ID = id
		case id:
		default:
			return nil, fmt.Errorf(
				"ambiguous maddr %s could refer to %s or %s",
				maddr,
				info.ID,
				id,
			)
		}
		info.Addrs = append(info.Addrs, taddr)
	}
	return &info, nil
}

func ParseAndResolveRdvpMaddrs(ctx context.Context, log *zap.Logger, addrs []string) ([]*peer.AddrInfo, error) {
	// Resolve all addresses
	outPeersUnmatched := make([]*peer.AddrInfo, len(addrs))
	var errs error
	var outLock sync.Mutex
	var wg sync.WaitGroup
	wg.Add(len(addrs))
	for i, v := range addrs {
		go func(j int, addr string) {
			defer wg.Done()

			rdvpeer, err := ParseAndResolveIpfsAddr(ctx, addr)
			if err != nil {
				outLock.Lock()
				defer outLock.Unlock()
				errs = multierr.Append(errs, err)
				return
			}

			fds := make([]zapcore.Field, len(rdvpeer.Addrs))
			for i, maddr := range rdvpeer.Addrs {
				key := fmt.Sprintf("#%d", i)
				fds[i] = zap.String(key, maddr.String())
			}
			log.Debug("rdvp peer resolved addrs", fds...)
			outPeersUnmatched[j] = rdvpeer
		}(i, v)
	}
	wg.Wait()
	if errs != nil {
		return nil, errs
	}
	// Match peers by ID
	outPeersMatched := make(map[peer.ID][]ma.Multiaddr)
	for _, v := range outPeersUnmatched {
		outPeersMatched[v.ID] = append(outPeersMatched[v.ID], v.Addrs...)
	}

	// Create the ultimate *peer.AddrInfo
	var outPeers []*peer.AddrInfo
	for id, maddrs := range outPeersMatched {
		outPeers = append(outPeers, &peer.AddrInfo{
			ID:    id,
			Addrs: maddrs,
		})
	}

	return outPeers, nil
}

var ErrExpectedEOF = errors.New("red data when expecting EOF")

const DefaultCloseTimeout = time.Second * 5

func FullClose(s network.Stream) error {
	// Start the close.
	err := s.CloseWrite()
	if err != nil {
		return err
	}

	// We don't want to wait indefinitely
	_ = s.SetDeadline(time.Now().Add(DefaultCloseTimeout))

	// Trying with a long slice to fetch `n`.
	n, err := s.Read([]byte{0})
	if n > 0 || err == nil {
		_ = s.Reset()
		return ErrExpectedEOF
	}
	if err == io.EOF {
		return nil
	}
	_ = s.Reset()
	return err
}
