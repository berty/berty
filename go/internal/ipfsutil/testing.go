package ipfsutil

import (
	"context"
	crand "crypto/rand"
	"encoding/base64"
	"testing"

	tinder "berty.tech/berty/v2/go/internal/tinder"
	ds "github.com/ipfs/go-datastore"
	dsync "github.com/ipfs/go-datastore/sync"
	"github.com/stretchr/testify/require"
	"go.uber.org/zap"

	ipfs_cfg "github.com/ipfs/go-ipfs-config"
	ipfs_core "github.com/ipfs/go-ipfs/core"
	ipfs_coreapi "github.com/ipfs/go-ipfs/core/coreapi"
	ipfs_mock "github.com/ipfs/go-ipfs/core/mock"
	ipfs_repo "github.com/ipfs/go-ipfs/repo"
	ipfs_interface "github.com/ipfs/interface-go-ipfs-core"

	libp2p_ci "github.com/libp2p/go-libp2p-core/crypto"
	host "github.com/libp2p/go-libp2p-core/host"
	p2pnetwork "github.com/libp2p/go-libp2p-core/network"
	"github.com/libp2p/go-libp2p-core/peer"
	libp2p_peer "github.com/libp2p/go-libp2p-core/peer"
	"github.com/libp2p/go-libp2p-core/protocol"
	rendezvous "github.com/libp2p/go-libp2p-rendezvous"
	p2p_rpdb "github.com/libp2p/go-libp2p-rendezvous/db/sqlite"
	libp2p_mocknet "github.com/libp2p/go-libp2p/p2p/net/mock"
)

// CoreAPIMock implements ipfs.CoreAPI and adds some debugging helpers
type CoreAPIMock interface {
	ExtendedCoreAPI

	Tinder() tinder.Driver
	MockNetwork() libp2p_mocknet.Mocknet
	MockNode() *ipfs_core.IpfsNode
	Close()
}

func TestingRepo(t testing.TB) ipfs_repo.Repo {
	t.Helper()

	c := ipfs_cfg.Config{}
	priv, pub, err := libp2p_ci.GenerateKeyPairWithReader(libp2p_ci.RSA, 2048, crand.Reader)
	if err != nil {
		t.Fatalf("failed to generate pair key: %v", err)
	}

	pid, err := libp2p_peer.IDFromPublicKey(pub)
	if err != nil {
		t.Fatalf("failed to get pid from pub key: %v", err)
	}

	privkeyb, err := priv.Bytes()
	if err != nil {
		t.Fatalf("failed to get raw priv key: %v", err)
	}

	c.Bootstrap = []string{}
	c.Addresses.Swarm = []string{"/ip6/::/tcp/0"}
	c.Identity.PeerID = pid.Pretty()
	c.Identity.PrivKey = base64.StdEncoding.EncodeToString(privkeyb)

	dstore := dsync.MutexWrap(ds.NewMapDatastore())
	return &ipfs_repo.Mock{
		D: dstore,
		C: c,
	}
}

type TestingAPIOpts struct {
	Logger  *zap.Logger
	Mocknet libp2p_mocknet.Mocknet
	RDVPeer peer.AddrInfo
}

// TestingCoreAPIUsingMockNet returns a fully initialized mocked Core API with the given mocknet
func TestingCoreAPIUsingMockNet(ctx context.Context, t testing.TB, opts *TestingAPIOpts) (api CoreAPIMock, cleanup func()) {
	t.Helper()

	if opts.Logger == nil {
		opts.Logger = zap.NewNop()
	}

	r := TestingRepo(t)
	routingopt, crout := NewTinderRouting(opts.Logger, &opts.RDVPeer, false)

	node, err := ipfs_core.NewNode(ctx, &ipfs_core.BuildCfg{
		Repo:    r,
		Online:  true,
		Routing: routingopt,
		Host:    ipfs_mock.MockHostOption(opts.Mocknet),
		ExtraOpts: map[string]bool{
			"pubsub": true,
		},
	})
	require.NoError(t, err, "failed to initialize IPFS node mock")

	// get back tinder service
	routing := <-crout

	coreapi, err := ipfs_coreapi.NewCoreAPI(node)
	require.NoError(t, err, "failed to initialize IPFS Core API mock")

	api = &coreAPIMock{
		CoreAPI: coreapi,
		mocknet: opts.Mocknet,
		node:    node,
		tinder:  routing.Routing,
	}

	cleanup = func() {
		node.Close()

		// manually close dht since ipfs will not be able to do that
		routing.IpfsDHT.Close()
	}

	return
}

// TestingCoreAPI returns a fully initialized mocked Core API.
// If you want to do some tests involving multiple peers you should use
// `TestingCoreAPIUsingMockNet` with the same mocknet instead.
func TestingCoreAPI(ctx context.Context, t testing.TB) (CoreAPIMock, func()) {
	t.Helper()

	m := libp2p_mocknet.New(ctx)
	defer func() {
		_ = m.LinkAll()
		_ = m.ConnectAllButSelf()
	}()

	peer, err := m.GenPeer()
	require.NoError(t, err)

	_, cleanrdvp := TestingRDVP(ctx, t, peer)
	api, cleanapi := TestingCoreAPIUsingMockNet(ctx, t, &TestingAPIOpts{
		Mocknet: m,
		RDVPeer: peer.Network().Peerstore().PeerInfo(peer.ID()),
	})

	cleanup := func() {
		cleanapi()
		cleanrdvp()
	}
	return api, cleanup
}

func TestingRDVP(ctx context.Context, t testing.TB, h host.Host) (*rendezvous.RendezvousService, func()) {
	db, err := p2p_rpdb.OpenDB(ctx, ":memory:")
	require.NoError(t, err)

	svc := rendezvous.NewRendezvousService(h, db)
	cleanup := func() {
		_ = db.Close()
	}
	return svc, cleanup
}

type coreAPIMock struct {
	ipfs_interface.CoreAPI

	mocknet libp2p_mocknet.Mocknet
	node    *ipfs_core.IpfsNode
	tinder  tinder.Driver
}

func (m *coreAPIMock) ConnMgr() ConnMgr {
	return m.node.PeerHost.ConnManager()
}

func (m *coreAPIMock) NewStream(ctx context.Context, p libp2p_peer.ID, pids ...protocol.ID) (p2pnetwork.Stream, error) {
	return m.node.PeerHost.NewStream(ctx, p, pids...)
}

func (m *coreAPIMock) SetStreamHandler(pid protocol.ID, handler p2pnetwork.StreamHandler) {
	m.node.PeerHost.SetStreamHandler(pid, handler)
}

func (m *coreAPIMock) RemoveStreamHandler(pid protocol.ID) {
	m.node.PeerHost.RemoveStreamHandler(pid)
}

func (m *coreAPIMock) MockNetwork() libp2p_mocknet.Mocknet {
	return m.mocknet
}

func (m *coreAPIMock) MockNode() *ipfs_core.IpfsNode {
	return m.node
}

func (m *coreAPIMock) Tinder() tinder.Driver {
	return m.tinder
}

func (m *coreAPIMock) Close() {
	m.node.Close()
}
