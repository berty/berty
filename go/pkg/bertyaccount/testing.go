package bertyaccount

import (
	"context"
	"testing"

	"github.com/stretchr/testify/assert"
	"google.golang.org/grpc"

	"berty.tech/berty/v2/go/internal/grpcutil"
	"berty.tech/berty/v2/go/pkg/accounttypes"
)

func TestingAccountClient(ctx context.Context, t *testing.T, s accounttypes.AccountServiceServer) accounttypes.AccountServiceClient {
	t.Helper()

	srv := grpc.NewServer()
	accounttypes.RegisterAccountServiceServer(srv, s)

	l := grpcutil.NewBufListener(ctx, 2048)

	cc, err := l.NewClientConn()
	assert.NoError(t, err)

	cl := accounttypes.NewAccountServiceClient(cc)

	go srv.Serve(l.Listener)

	t.Cleanup(func() { l.Close() })

	return cl
}
