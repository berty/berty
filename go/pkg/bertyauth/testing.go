package bertyauth

import (
	"crypto/ed25519"
	crand "crypto/rand"
	"testing"

	"github.com/stretchr/testify/require"

	"berty.tech/berty/v2/go/pkg/cryptoutil"
)

func HelperGenerateTokenIssuerSecrets(t *testing.T) ([]byte, ed25519.PublicKey, ed25519.PrivateKey) {
	t.Helper()

	pk, sk, err := ed25519.GenerateKey(crand.Reader)
	require.NoError(t, err)

	secret := make([]byte, cryptoutil.KeySize)
	_, err = crand.Read(secret)
	require.NoError(t, err)

	return secret, pk, sk
}
