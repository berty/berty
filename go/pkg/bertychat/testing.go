package bertychat

import (
	"testing"

	"berty.tech/go/internal/chatdb"
	"berty.tech/go/pkg/bertyprotocol"
	"go.uber.org/zap"
)

// TestingClient returns a configured Client struct with in-memory contexts.
func TestingClient(t *testing.T, opts Opts) (Client, func()) {
	t.Helper()

	if opts.Logger == nil {
		opts.Logger = zap.NewNop()
	}

	db := chatdb.TestingSqliteDB(t, opts.Logger)
	protocol, protocolCleanup := bertyprotocol.TestingClient(t, bertyprotocol.Opts{Logger: opts.Logger})

	client, err := New(db, protocol, opts)
	if err != nil {
		t.Fatalf("failed to initialize client: %v", err)
	}

	cleanup := func() {
		client.Close()
		db.Close()
		protocolCleanup()
	}

	return client, cleanup
}
