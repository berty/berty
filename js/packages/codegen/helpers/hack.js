const Case = require('case')
module.exports.register = (handlebars) => {
	// try to match to string that are inconsistants
	handlebars.registerHelper('matchInconsistant', function (a, b) {
		// todo: map here
		if (b.replace('Outgoing', '') === a) {
			return true
		}
		if (b.replace('Incoming', '') === a) {
			return true
		}
		const _a = Case.snake(a)
		const _b = Case.snake(b)
		const _aSplited = _a.split('_')
		const _bSplited = _b.split('_')
		let ratio = _aSplited.length > _bSplited.length ? -_aSplited.length : -_bSplited.length
		_a.split('_').forEach((__a) => {
			if (Case.pascal(b).includes(Case.pascal(__a))) {
				ratio += 1
			}
		})
		if (ratio === 0) {
			return true
		}
		return false
	})
}
