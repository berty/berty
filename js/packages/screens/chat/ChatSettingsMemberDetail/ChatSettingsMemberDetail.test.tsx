import { getFirstMultiMemberConv } from '@berty/utils/testing/helpers'
import { mockServices } from '@berty/utils/testing/mockServices.test'
import { renderScreen } from '@berty/utils/testing/renderScreen.test'

import { ChatSettingsMemberDetail } from './ChatSettingsMemberDetail'

test('Group.ChatSettingsMemberDetail renders correctly', async () => {
	await mockServices()

	const conv = getFirstMultiMemberConv()

	const { toJSON } = renderScreen('Group.ChatSettingsMemberDetail', ChatSettingsMemberDetail, {
		convId: conv?.publicKey || '',
		memberPk: conv?.members ? conv.members[0].publicKey || '' : '',
		displayName: conv?.members ? conv.members[0].displayName || '' : '',
	})
	expect(toJSON()).toMatchSnapshot()
})
