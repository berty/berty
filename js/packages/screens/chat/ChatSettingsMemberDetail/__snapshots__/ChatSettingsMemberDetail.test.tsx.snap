// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Group.ChatSettingsMemberDetail renders correctly 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  />
</RNCSafeAreaProvider>
`;
